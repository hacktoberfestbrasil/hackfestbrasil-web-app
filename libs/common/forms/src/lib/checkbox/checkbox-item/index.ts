export * from './checkbox-item.component';
