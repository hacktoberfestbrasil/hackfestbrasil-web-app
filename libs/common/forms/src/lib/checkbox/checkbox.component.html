<div class="hackfest-checkbox">
  <input
    #input
    [id]="id"
    name="checks"
    type="checkbox"
    class="cdk-visually-hidden"
    [value]="value"
    [checked]="el?.checked"
    [attr.aria-checked]="el?.checked"
    [formControl]="control"
    (change)="onChangeEvent($event)"
  />
  <label [attr.for]="id">
    <ng-content></ng-content>

    <svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg">
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
