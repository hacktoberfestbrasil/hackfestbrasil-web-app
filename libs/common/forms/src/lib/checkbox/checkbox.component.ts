import {
  Self,
  Input,
  Output,
  Optional,
  ViewChild,
  Component,
  Injectable,
  ElementRef,
  forwardRef,
  OnDestroy,
  EventEmitter,
  AfterViewInit,
  AfterContentInit,
  ChangeDetectionStrategy,
} from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { EventInputTarget } from './../types';
import { HackfestControlAccessor } from '../abstractions';
import {
  AbstractControl,
  FormControl,
  NgControl,
  NG_VALUE_ACCESSOR,
} from '@angular/forms';

let nextId = 0;

@Injectable()
export class HackfestCheckbox extends HackfestControlAccessor {
  @Input()
  public set disabled(value: boolean) {
    this._disabled = value;
  }
}

const HackfestCheckboxProvider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => HackfestCheckbox),
  multi: true,
};

@Component({
  selector: 'hackfest-checkbox',
  templateUrl: './checkbox.component.html',
  styleUrls: ['./checkbox.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [HackfestCheckbox, HackfestCheckboxProvider],
})
export class HackfestCheckboxComponent extends HackfestCheckbox
  implements AfterContentInit, AfterViewInit, OnDestroy {
  destroy$ = new Subject<void>();

  @ViewChild('input', { static: true }) _el: ElementRef<HTMLInputElement>;
  get el() {
    return this._el.nativeElement;
  }

  @Input()
  public set id(value: string) {
    this._id = value;
  }
  public get id(): string {
    return this._id;
  }
  private _id = `hackfest-checkbox-${nextId++}`;


  @Input()
  public set value(value: any) {
    this._value = value;
  }
  public get value() {
    return this._value;
  }

  @Output()
  valueChange: EventEmitter<any> = new EventEmitter<any>();

  @Output()
  checkedChange = new EventEmitter<boolean>(false);

  control: AbstractControl;

  constructor(@Optional() @Self() public ngControl: NgControl) {
    super();
  }

  ngAfterContentInit() {
    this.control = this.ngControl?.control
      ? this.ngControl.control
      : new FormControl();
  }
  ngAfterViewInit() {
    this.ngControl?.control?.valueChanges
      .pipe(takeUntil(this.destroy$))
      .subscribe((value) => this.valueChange.emit(value));
  }

  onChangeEvent(evt: EventInputTarget) {
    this.onChange(evt.target.value);
    this.checkedChange.emit(evt.target.checked);
    this.valueChange.emit(evt.target.value);
  }

  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }
}
