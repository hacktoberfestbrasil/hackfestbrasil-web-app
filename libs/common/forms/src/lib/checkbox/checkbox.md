## Import

```ts
import { HackfestCheckboxModule } from '@hackfest/forms';
```


