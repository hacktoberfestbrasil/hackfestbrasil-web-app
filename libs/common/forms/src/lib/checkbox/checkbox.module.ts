import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HackfestCheckboxComponent } from './checkbox.component';
import { HackfestCheckboxGroupComponent } from './checkbox-group/checkbox-group.component';
import { HackfestCheckboxItemComponent } from './checkbox-item/checkbox-item.component';

export * from './checkbox.component';
export * from './checkbox-item/checkbox-item.component';
export * from './checkbox-group/checkbox-group.component';

@NgModule({
  imports: [FormsModule, CommonModule, ReactiveFormsModule],
  declarations: [
    HackfestCheckboxComponent,
    HackfestCheckboxGroupComponent,
    HackfestCheckboxItemComponent,
  ],
  exports: [
    HackfestCheckboxComponent,
    HackfestCheckboxGroupComponent,
    HackfestCheckboxItemComponent,
  ],
})
export class HackfestCheckboxModule {}
