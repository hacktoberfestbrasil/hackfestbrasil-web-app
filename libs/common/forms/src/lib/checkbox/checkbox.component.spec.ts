import { Component } from '@angular/core';
import { FormsModule, ReactiveFormsModule, FormControl } from '@angular/forms';
import {
  Spectator,
  SpectatorHost,
  createHostFactory,
  createComponentFactory,
} from '@ngneat/spectator';

import { HackfestCheckboxComponent } from './checkbox.component';

const VIEWS = {
  onlyCheck: `<hackfest-checkbox [formControl]="form"></hackfest-checkbox>`,
  withLabel: `<hackfest-checkbox [formControl]="form">Label</hackfest-checkbox>`,
  withCustomId: `<hackfest-checkbox id="x" [formControl]="form"></hackfest-checkbox>`,
};

@Component({ selector: 'hackfest-custom-form', template: '' })
class CustomFormComponent {
  form = new FormControl();
}

describe('HackfestCheckboxComponent', () => {
  describe('Host', () => {
    let spectator: SpectatorHost<HackfestCheckboxComponent, CustomFormComponent>;
    const createHost = createHostFactory({
      component: HackfestCheckboxComponent,
      host: CustomFormComponent,
      imports: [ReactiveFormsModule],
    });

    describe('Views', () => {
      it('should display only input', () => {
        spectator = createHost(VIEWS.onlyCheck);
        expect(spectator.query('.hackfest-checkbox')).toMatchSnapshot();
      });

      it('should display checkbox with label', () => {
        spectator = createHost(VIEWS.withLabel);
        expect(spectator.query('.hackfest-checkbox')).toMatchSnapshot();
      });

      it('should display checkbox with custom id', () => {
        spectator = createHost(VIEWS.withCustomId);
        expect(spectator.query('.hackfest-checkbox')).toMatchSnapshot();
      });
    });

    describe('Props', () => {
      it('should create an instance', () => {
        spectator = createHost(VIEWS.withCustomId);
        const input = spectator.query('.hackfest-checkbox > input');
        expect(input.getAttribute('id')).toEqual('x');
      });
    });
  });

  describe('Forms', () => {
    let spectator: Spectator<HackfestCheckboxComponent>;
    const createComponent = createComponentFactory({
      component: HackfestCheckboxComponent,
      imports: [FormsModule, ReactiveFormsModule],
    });

    it('should create', () => {
      spectator = createComponent();
      expect(spectator.component).toBeTruthy();
    });

    it('should value changed if set value', () => {
      spectator.component.control.setValue('abc');
      expect(spectator.component.control.value).toBe('abc');
    });

    it('should value changed if set value', async () => {
      spyOn(spectator.component.valueChange, 'emit');
      spectator.component.control.valueChanges.subscribe(() => {
        expect(spectator.component.onChangeEvent).toBeCalledTimes(1);
        expect(spectator.component.valueChange.emit).toBeCalledTimes(1);
        expect(spectator.component.checkedChange.emit).toBeCalledTimes(1);
      });
      spectator.component.control.setValue('qwe');
    });
  });
});

