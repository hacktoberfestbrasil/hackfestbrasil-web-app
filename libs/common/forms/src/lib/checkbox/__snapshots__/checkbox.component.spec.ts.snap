// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackfestCheckboxComponent Host Views should display checkbox with custom id 1`] = `
<div
  class="hackfest-checkbox"
>
  <input
    aria-checked="false"
    class="cdk-visually-hidden ng-untouched ng-pristine ng-valid"
    id="x"
    name="checks"
    type="checkbox"
    value="undefined"
  />
  <label
    for="x"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;

exports[`HackfestCheckboxComponent Host Views should display checkbox with label 1`] = `
<div
  class="hackfest-checkbox"
>
  <input
    aria-checked="false"
    class="cdk-visually-hidden ng-untouched ng-pristine ng-valid"
    id="hackfest-checkbox-1"
    name="checks"
    type="checkbox"
    value="undefined"
  />
  <label
    for="hackfest-checkbox-1"
  >
    Label
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;

exports[`HackfestCheckboxComponent Host Views should display only input 1`] = `
<div
  class="hackfest-checkbox"
>
  <input
    aria-checked="false"
    class="cdk-visually-hidden ng-untouched ng-pristine ng-valid"
    id="hackfest-checkbox-0"
    name="checks"
    type="checkbox"
    value="undefined"
  />
  <label
    for="hackfest-checkbox-0"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22.3,61.3c5.7,4.1,13.2,11.8,16.7,18C53,60.5,66,41,77.7,20.8"
      />
    </svg>
  </label>
</div>
`;
